### PR TITLE
Create SNS Lambda

### PR DESCRIPTION
**Pull request by blakemorgan**
Add the Lambda code and infrastructure. This will create the Lambda function with the appropriate permissions and SNS topic subscriptions. The user just needs to provide the SNS Topic ARN, the messages from that ARN will be pushed ot Slack.

This has been tested in the new version of Payment Manager.

Screen Shot 2021-07-30 at 10 25 39 AM

### CODE REVIEW

File: README.md
Comment (yoshutch) [inline README.md]: (nit) would the error not occur if you created the sns topic in the "setup" phase before your app terraform config/apply? I wonder if we should recommend that or at least put that in here as a workaround so that you don't get this error?

**Review by yoshutch — COMMENTED** [review]
Does this add any formatting to the message before sending to slack? Or is it basically whatever you put into the SNS topic will be sent to slack?

File: README.md
Comment (blakemorgan) [reply README.md]: Oh yeah that will work. I copied this from the Humio logger module and that wasn't an option there because it depended on the log groups created from fargate, which depended on the setup resources. I'll change that.

**Comment by blakemorgan** [comment]
No formatting is added to the messages by this lambda other than prepending the `app_name` before the message. However, the user can format the message using markdown. SNS events have a `Message` property, and whatever is in there will be sent to Slack,

File: lambda/index.js
Comment (GaryGSC) [inline lambda/index.js:19]:
```suggestion
exports.handler = async function (event, context) {
  console.debug('Event: ' + JSON.stringify(event, null, 2))

  const webhookUrl = new URL(process.env.SLACK_WEBHOOK_URL)
  try {
    await Promise.all(event.Records.map(record => _sendSlackMessage(record.Sns.Message, webhookUrl)))
    console.info('All messages sent successfully.')
  } catch (e) {
    console.error(e)
    context.fail(e)
  }
}
```

File: lambda/index.js
Comment (GaryGSC) [inline lambda/index.js:25]:
```suggestion
// https://api.slack.com/reference/surfaces/formatting#escaping
const ampRegex = /&/g
const ltRegex = /</g
const gtRegex = />/g
function escapeForSlack (string) {
  return string.replace(ampRegex, '&amp;').replace(ltRegex, '&lt;').replace(gtRegex, '&gt;')
}

function _sendSlackMessage (messageText, webhookUrl) {
  const data = JSON.stringify({
    type: 'mrkdwn',
    text: `*[${process.env.APP_NAME.toUpperCase()}]* ${escapeForSlack(messageText)}`
  })
```

File: README.md
Comment (yoshutch) [inline README.md]: (nit) typo
```suggestion
Because of [limitations with Terraform](https://www.terraform.io/docs/language/meta-arguments/for_each.html#limitations-on-values-used-in-for_each), SNS Topics you want integrating with this module should be created in your one-time setup resources. You can then include this module in your app resources as normal. If that's not possible, you must add this module to your app resources after the initial deployment of your application to a new environment and creation of the SNS Topic it will integrate with. Otherwise, you'll get an error similar to this:
```

File: lambda/index.js
Comment (GaryGSC) [reply lambda/index.js:25]: I'm sure there's more we can do to pretty up the message, but this change will at least keep us from breaking our formatting.

**Review by yoshutch — APPROVED** [review]

File: lambda/index.js
Comment (GaryGSC) [inline lambda/index.js:39]: We could probably move this check earlier to save some processing time.

**Review by GaryGSC — COMMENTED** [review]
A few suggestions. Generally LGTM.

**Comment by GaryGSC** [comment]
We're fast with these merges. 😆

**Comment by blakemorgan** [comment]
Yeah sorry. I didn't know anyone else was looking.

**Comment by GaryGSC** [comment]
You're good. I tend to merge right after I get an approval or two, too.

**Pull request merged**
